src/types/special.rs

```diff
@@ -0,0 +1,129 @@
+use std::io::prelude::*;
+use std::io::{Cursor, SeekFrom};
+use std::{i32, i64};
+
+use byteorder::{ReadBytesExt, WriteBytesExt, BigEndian};
+
+use Result;
+use error::Error;
+use types::{Type, FromSql, ToSql, IsNull, SessionInfo};
+
+/// A wrapper that can be used to represent infinity with `Type::Date` types.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum Date<T> {
+    /// Represents `infinity`, a date that is later than all other dates.
+    PosInfinity,
+    /// Represents `-infinity`, a date that is earlier than all other dates.
+    NegInfinity,
+    /// The wrapped date.
+    Value(T),
+}
+
+impl<T: FromSql> FromSql for Date<T> {
+    fn from_sql<R: Read>(ty: &Type, raw: &mut R, ctx: &SessionInfo) -> Result<Self> {
+        if *ty != Type::Date {
+            return Err(Error::Conversion("expected date type".into()));
+        }
+
+        let mut buf = [0; 4];
+        try!(raw.read_exact(&mut buf));
+        let mut cursor = Cursor::new(&mut buf);
+
+        match try!(cursor.read_i32::<BigEndian>()) {
+            i32::MAX => Ok(Date::PosInfinity),
+            i32::MIN => Ok(Date::NegInfinity),
+            _ => {
+                try!(cursor.seek(SeekFrom::Start(0)));
```

Comment:
This can be replaced with just `cursor.set_position(0)`, but I think we can avoid the `Cursor` entirely by using `&mut [u8]`s `Read` implementation:

https://is.gd/XIpI1U